### PR TITLE
Add proper hand models during hand-tracking

### PR DESCRIPTION
**Pull request by elima**
This series replaces the current dummy hands by a rigged hand model on all devices but Pico. On Pico, the hand-tracking data we get from the runtime  is not accurate or stable enough to properly animate the models we currently have available, which are the ones provided by Immersive Web's work group: https://github.com/immersive-web/webxr-input-profiles/tree/main/packages/assets/profiles/generic-hand.

Due to existing limitation in the VRB library, we added a new class `HandModels` which encapsulates all GL(ES) code necessary for rendering the hands, including the shaders. Follow-up work will extend VRB to support geometry and shaders that animates rigged geometry, then we could integrate and remove this class.

The material and lighting used for the hand models is currently hard-coded. Follow-up work could make the appearance of the hands changeable by users e.g via settings.

The series has been tested extensively on Pico4 and Quest2. The former keeps the current hands representation based on a sphere for each joint, as mentioned above. On Quest2 the new hand models perform pretty good.

### CODE REVIEW

**Comment by felipeerias** [comment]
The new hand model works very well in the Oculus, good job :slightly_smiling_face: 

However, the spheres in Pico look worse than before. I don't know where that line in the middle of each sphere comes from.

Screenshot_20230518_123432

(I was also wondering if we could make the hand a bit more interesting by changing the colors used in the light and material, but it is better to leave that for a different PR).

**Comment by felipeerias** [comment]
@elima I suspect that the problem comes from the sphere geometry that we get from `DeviceUtils::GetSphereGeometry`.

**Comment by felipeerias** [comment]
@elima PR https://github.com/Igalia/wolvic/pull/677 fixes the issue with the spheres.

File: .gitmodules [hunk 1]
```diff
@@ -1,3 +1,7 @@
 [submodule "app/src/main/cpp/vrb"]
 	path = app/src/main/cpp/vrb
 	url = https://github.com/igalia/vrb.git
+[submodule "app/src/main/cpp/tinygltf"]
+	path = app/src/main/cpp/tinygltf
+	url = https://github.com/Igalia/tinygltf.git
```
Comment (svillar) [inline .gitmodules:6]: Hmm why are we using a fork instead of pointing to the original repo?

File: app/CMakeLists.txt [hunk 2]
```diff
@@ -220,6 +220,20 @@ if (SPACES)
...
 endif()
 
+
+# Add dependency on tinygltf library, which is used to load hand models
+# from .glb assets. Since it is only a single C++ source and header,
+# we include them as native-lib's own build sources.
+add_definitions(-DTINYGLTF_ANDROID_LOAD_FROM_ASSETS)
+target_sources(
+    native-lib
+    PUBLIC
```
Comment (svillar) [inline app/CMakeLists.txt:231]: Nit let's make this a bit more compact, we can put everything but the path in the same line

File: app/src/main/cpp/Controller.h [hunk 3]
```diff
@@ -45,6 +45,9 @@ struct Controller {
   vrb::TogglePtr modelToggle;
   vrb::TogglePtr handMeshToggle;
   std::vector<vrb::TransformPtr> handJointTransforms;
+  ControllerDelegate::HandMesh handMesh;
+  bool hasHandMesh;
```
Comment (svillar) [inline app/src/main/cpp/Controller.h:49]: Why don't we store the hand mesh in an `unique_ptr` and create it on demand?

It has also the advantage of not having to add the extra `hasHandMesh` because you could perform null checks

File: app/src/main/cpp/ControllerDelegate.h [hunk 4]
```diff
@@ -32,6 +35,17 @@ enum class HandTrackingJoints {
   RingMetacarpal, RingProximal, RingIntermediate, RingDistal, RingTip,
   LittleMetacarpal, LittleProximal, LittleIntermediate, LittleDistal, LittleTip };
 
+struct Vector2f {
+  float x;
+  float y;
+};
```
Comment (svillar) [inline app/src/main/cpp/ControllerDelegate.h:41]: Seems like this is unused?

File: app/src/main/cpp/HandModels.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
+/* -*- Mode: C++; tab-width: 20; indent-tabs-mode: nil; c-basic-offset: 2 -*-
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef VRBROWSER_HANDMODELS_H
+#define VRBROWSER_HANDMODELS_H
```
Comment (svillar) [inline app/src/main/cpp/HandModels.h:7]: Instead of doing this (which is old fashioned) please simply use `#pragma once`

File: app/src/main/cpp/HandModels.cpp [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+  State()
+      : vertexShader(0)
+      , fragmentShader(0)
+      , program(0)
+      , aPosition(0)
+      , aNormal(0)
+      , aJointIndices(0)
+      , aJointWeights(0)
+      , handGLState({})
+  {}
```
Comment (svillar) [inline app/src/main/cpp/HandModels.cpp:157]: Nitpick, I do prefer to initialize the attributes in the declaration
```
GLuint vertexShader { 0 };
GLuing fragmentShader { 0 };
...
````
And then you can just use the default constructor.

File: app/src/main/cpp/HandModels.cpp [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+};
+
+HandModelsPtr
+HandModels::Create(vrb::CreationContextPtr& aContext) {
+  return std::make_shared<vrb::ConcreteClass<HandModels, HandModels::State> >(aContext);
+}
+
+void
+HandModels::StartFrame() {
+}
```
Comment (svillar) [inline app/src/main/cpp/HandModels.cpp:167]: Perhaps add the `{}` in the declaration and we don't need to add this empty implementation in the cpp file

File: app/src/main/cpp/HandModels.cpp [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+
+void
+HandModels::Draw(const vrb::Camera& aCamera, const Controller& aController) {
+  if (!aController.hasHandMesh)
+    return;
+
+  if (!m.program) {
+    VRB_ERROR("HandModel::Draw FAILED!");
+    return;
+  }
```
Comment (svillar) [inline app/src/main/cpp/HandModels.cpp:177]: I think this should be instead an assert right? Because I don't think any normal flow could lead to that

File: app/src/main/cpp/HandModels.cpp [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+  if (!enabled)
+    VRB_GL_CHECK(glEnable(GL_DEPTH_TEST));
+
+  glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+  glEnable(GL_BLEND);
+
+  VRB_GL_CHECK(glUniformMatrix4fv(m.uPerspective, 1, GL_FALSE, aCamera.GetPerspective().Data()));
+  VRB_GL_CHECK(glUniformMatrix4fv(m.uView, 1, GL_FALSE, aCamera.GetView().Data()));
+  // @FIXME: We don't apply any model transform for now, but eventually we could
+  // use `controller.transformMatrix`, if there is need.
```
Comment (svillar) [inline app/src/main/cpp/HandModels.cpp:193]: I don't exactly get this, which kind of transformation are we not applying?

File: app/src/main/cpp/HandModels.cpp [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+  assert(aController.meshJointTransforms.size() == state.jointCount);
+  std::vector<vrb::Matrix> jointMatrices;
+  jointMatrices.resize(state.jointCount);
+  for (int i = 0; i < state.jointCount; i++) {
+    jointMatrices[i] = aController.meshJointTransforms[i].PostMultiply(state.bindMatrices[i]);
+    VRB_GL_CHECK(glUniformMatrix4fv(m.uJointMatrices + i, 1, GL_FALSE, jointMatrices[i].Data()));
+  }
+
+  VRB_GL_CHECK(glBindBuffer(GL_ARRAY_BUFFER, state.vboPosition));
+  VRB_GL_CHECK(glEnableVertexAttribArray((GLuint)m.aPosition));
```
Comment (svillar) [inline app/src/main/cpp/HandModels.cpp:206]: Nit: as a general comment, do separate the type casting from the variable with a white space

File: app/src/main/cpp/HandModels.cpp [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+
+  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
+  glBindBuffer(GL_ARRAY_BUFFER, 0);
+
+  if (!enabled)
+    VRB_GL_CHECK(glDisable(GL_DEPTH_TEST));
+}
+
+void HandModels::UpdateHandModel(const Controller& aController) {
+  VRB_LOG("HandModel: %u vertices, %u indices", aController.handMesh.vertexCount, aController.handMesh.indexCount);
```
Comment (svillar) [inline app/src/main/cpp/HandModels.cpp:232]: Not sure we need a logging statement here. In any case if you want to keep it, it should be more descriptive, perhaps adding the method name?

File: app/src/main/cpp/HandModels.cpp [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+  state.jointCount = mesh.jointCount;
+  state.vertexCount = mesh.vertexCount;
+  state.indexCount = mesh.indexCount;
+
+  if (state.iboIndices == 0) {
+    VRB_GL_CHECK(glGenBuffers(1, &state.vboPosition));
+    VRB_GL_CHECK(glGenBuffers(1, &state.vboNormal));
+    VRB_GL_CHECK(glGenBuffers(1, &state.vboJointIndices));
+    VRB_GL_CHECK(glGenBuffers(1, &state.vboJointWeights));
+
```
Comment (svillar) [inline app/src/main/cpp/HandModels.cpp:249]: maybe remove this empty line?

File: app/src/main/cpp/HandModels.cpp [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+  // Joint weights VBO
+  VRB_GL_CHECK(glBindBuffer(GL_ARRAY_BUFFER, state.vboJointWeights));
+  VRB_GL_CHECK(glBufferData(GL_ARRAY_BUFFER, state.vertexCount * 4 * sizeof(float), mesh.jointWeights.data(), GL_STATIC_DRAW));
+
+  // Indices IBO
+  VRB_GL_CHECK(glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, state.iboIndices));
+  VRB_GL_CHECK(glBufferData(GL_ELEMENT_ARRAY_BUFFER, state.indexCount * 1 * sizeof(uint16_t), mesh.indices.data(), GL_STATIC_DRAW));
+
+  // Joint bind matrices
+  state.bindMatrices.resize(XR_EXT_HAND_TRACKING_NUM_JOINTS);
```
Comment (svillar) [inline app/src/main/cpp/HandModels.cpp:282]: Hmm that variable is from OpenXR, we should not use it here directly dince this code is in theory generic.

File: app/src/main/cpp/HandModels.cpp [hunk 14]
```diff
@@ -0,0 +1,356 @@
...
+  for (int i = 0; i < state.jointCount; i++)
+    state.bindMatrices[i] = mesh.jointTransforms[i];
+
+  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
+  glBindBuffer(GL_ARRAY_BUFFER, 0);
+}
+
+void
+HandModels::EndFrame() {
+}
```
Comment (svillar) [inline app/src/main/cpp/HandModels.cpp:292]: Ditto adding braces to definition and removing this.

File: app/src/main/cpp/HandModels.cpp [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+
+void
+HandModels::StopPresenting() {
+}
```
Comment (svillar) [inline app/src/main/cpp/HandModels.cpp:296]: Ditto

File: app/src/main/cpp/HandModels.cpp [hunk 16]
```diff
@@ -0,0 +1,356 @@
...
+
+void
+HandModels::CancelFrame(const int32_t aSurfaceHandle) {
+}
```
Comment (svillar) [inline app/src/main/cpp/HandModels.cpp:300]: Ditto

File: app/src/main/cpp/HandModels.cpp [hunk 17]
```diff
@@ -0,0 +1,356 @@
...
+{}
+
+void
+HandModels::InitializeGL() {
+  m.vertexShader = vrb::LoadShader(GL_VERTEX_SHADER, sVertexShader);
+  m.fragmentShader = vrb::LoadShader(GL_FRAGMENT_SHADER, sFragmentShader);
+  if (m.vertexShader && m.fragmentShader)
+    m.program = vrb::CreateProgram(m.vertexShader, m.fragmentShader);
+
+  if (!m.program)
```
Comment (svillar) [inline app/src/main/cpp/HandModels.cpp:314]: Should this be an assert?

File: app/src/main/cpp/BrowserWorld.cpp [hunk 18]
```diff
@@ -1134,6 +1152,7 @@ BrowserWorld::StartFrame() {
     if (relayoutWidgets) {
       UpdateVisibleWidgets();
     }
+    m.handModelsRenderer->StartFrame();
```
Comment (svillar) [inline app/src/main/cpp/BrowserWorld.cpp:1155]: I wonder whether we can get rid of these knowing that we have empty implementations.

File: app/src/main/cpp/BrowserWorld.cpp [hunk 19]
```diff
@@ -1151,6 +1170,8 @@ BrowserWorld::EndFrame() {
   }
   m.drawHandler = nullptr;
 
+  m.handModelsRenderer->EndFrame();
```
Comment (svillar) [inline app/src/main/cpp/BrowserWorld.cpp:1173]: Ditto

File: app/src/main/cpp/BrowserWorld.cpp [hunk 20]
```diff
@@ -1696,6 +1717,12 @@ BrowserWorld::DrawWorld(device::Eye aEye) {
     m.drawList->Draw(*camera);
   }
 
+  // Draw hand models
+  for (Controller& controller: m.controllers->GetControllers()) {
+    if (controller.enabled && controller.mode == ControllerMode::Hand)
+      m.handModelsRenderer->Draw(*camera, controller);
+  }
```
Comment (svillar) [inline app/src/main/cpp/BrowserWorld.cpp:1724]: Hmm shouldn't we check that we have the mesh available? I wonder what happens in the case of Pico for example

File: app/src/main/cpp/ControllerContainer.cpp [hunk 21]
```diff
@@ -203,38 +204,42 @@ void ControllerContainer::SetHandJointLocations(const int32_t aControllerIndex,
...
-            icon->UpdateProgram("");
-
-            controller.handActionButtonToggle = Toggle::Create(create);
-            controller.handActionButtonTransform = Transform::Create(create);
-            controller.handActionButtonToggle->AddNode(controller.handActionButtonTransform);
-            controller.handActionButtonTransform->AddNode(icon->GetRoot());
-            controller.handActionButtonToggle->ToggleAll(false);
-            m.root->AddNode(controller.handActionButtonToggle);
-        }
+    if (controller.leftHanded && controller.handActionButtonToggle == nullptr) {
```
Comment (svillar) [inline app/src/main/cpp/ControllerContainer.cpp:213]: Let's keep the (wrong) indentation as it was in order not to pollute the already big change and unnecessarily alter the blame history of those lines

File: app/src/main/cpp/ControllerContainer.cpp [hunk 22]
```diff
@@ -203,38 +204,42 @@ void ControllerContainer::SetHandJointLocations(const int32_t aControllerIndex,
...
+        m.root->AddNode(controller.handActionButtonToggle);
+    }
 #endif
 
+    // Due to unstable or inaccurate hand-tracking data, hand mesh models
+    // are not enabled on Pico devices. We are drawing a sphere for each joint
+    // instead.
+#if defined(PICOXR)
+    // Initialize hand joints if needed
+    if (controller.handJointTransforms.size() == 0) {
```
Comment (svillar) [inline app/src/main/cpp/ControllerContainer.cpp:240]: Hmm why is this needed now? Looks like it might be a different bug?

File: app/src/main/cpp/ControllerContainer.cpp [hunk 23]
```diff
@@ -270,6 +275,25 @@ void ControllerContainer::SetHandJointLocations(const int32_t aControllerIndex,
...
+#else
+    std::vector<vrb::Matrix> tmpJointTransforms = jointTransforms;
+    // We ignore the first matrix, corresponding to the palm, because the
+    // models we are currently using don't include a palm joint.
+    tmpJointTransforms.erase(tmpJointTransforms.begin());
+
+    if (controller.meshJointTransforms.size() == 0)
+        controller.meshJointTransforms.resize(tmpJointTransforms.size());
+    else
+        assert(controller.meshJointTransforms.size() == tmpJointTransforms.size());
```
Comment (svillar) [inline app/src/main/cpp/ControllerContainer.cpp:287]: Just place the assert after the if and not in the else branch, because it should be true even if the if branch was taken.

File: app/src/main/cpp/ControllerContainer.cpp [hunk 24]
```diff
@@ -318,6 +342,130 @@ ControllerContainer::GetControllers() const {
...
+bool ControllerContainer::LoadHandMeshFromAssets(Controller& aController) {
+  auto& handMesh = aController.handMesh;
+
+  tinygltf::TinyGLTF modelLoader;
+  tinygltf::Model model;
+  std::string err;
+  std::string warn;
+  bool ok = modelLoader.LoadBinaryFromFile(&model, &err, &warn,
+                                           aController.leftHanded ? "hand-model-left.glb" : "hand-model-right.glb");
+  if (!ok) {
```
Comment (svillar) [inline app/src/main/cpp/ControllerContainer.cpp:377]: `ok` is not really needed so I think we can just do `if (!modelLoader....) { VRB_ERROR... ; return false }`

File: app/src/main/cpp/ControllerContainer.cpp [hunk 25]
```diff
@@ -318,6 +342,130 @@ ControllerContainer::GetControllers() const {
...
+    VRB_ERROR("Error loading hand mesh asset: %s", err.c_str());
+    return false;
+  }
+  if (!warn.empty()) {
+    VRB_WARN("%s", warn.c_str());
+  }
```
Comment (svillar) [inline app/src/main/cpp/ControllerContainer.cpp:383]: remove braces

File: app/src/main/cpp/ControllerContainer.cpp [hunk 26]
```diff
@@ -318,6 +342,130 @@ ControllerContainer::GetControllers() const {
...
+    return false;
+  }
+  handMesh.indexCount = handMesh.indices.size();
+
+  // Load vertex attributes
+  for (auto& attr: primitive.attributes) {
+    if (attr.second >= model.accessors.size())
+      return false;
+
+    bool error = false;
```
Comment (svillar) [inline app/src/main/cpp/ControllerContainer.cpp:410]: I do prefer using here the opposite of error and not negate the output of `loadHandMeshAttribute` calls which is a bit weird.

File: app/src/main/cpp/ControllerContainer.cpp [hunk 27]
```diff
@@ -318,6 +342,130 @@ ControllerContainer::GetControllers() const {
...
+      if (bufferView.buffer >= model.buffers.size())
+        return false;
+      auto& buffer = model.buffers[bufferView.buffer];
+      if (accessor.type != TINYGLTF_TYPE_VEC4 || accessor.componentType != TINYGLTF_COMPONENT_TYPE_UNSIGNED_BYTE)
+        return false;
+      if (bufferView.byteLength != accessor.count * sizeof(uint8_t) * 4)
+        return false;
+      handMesh.jointIndices.resize(accessor.count);
+      for (int i = 0; i < accessor.count; i++) {
+        unsigned char *ptr = buffer.data.data() + bufferView.byteOffset + accessor.byteOffset + i * 4;
```
Comment (svillar) [inline app/src/main/cpp/ControllerContainer.cpp:436]: Nit `unsigned char*`

File: app/src/main/cpp/ControllerContainer.cpp [hunk 28]
```diff
@@ -246,12 +246,12 @@ void ControllerContainer::SetHandJointLocations(const int32_t aControllerIndex,
...
+        vrb::Color handColor = vrb::Color(0.85f, 0.85f, 0.85f);
         handColor.SetAlpha(1.0);
-        state->SetMaterial(handColor, handColor,
-                           Color(0.0f, 0.0f, 0.0f),
-                           0.0f);
-        state->SetLightsEnabled(false);
+        state->SetMaterial(handColor, Color(0.5f, 0.5f, 0.5f),
+                           Color(0.5f, 0.5f, 0.5f),
+                           0.75f);
+        state->SetLightsEnabled(true);
```
Comment (svillar) [inline app/src/main/cpp/ControllerContainer.cpp:254]: @felipeerias is reporting that this is causing weird reflections in the spheres, maybe we don't need them.

**Review by svillar — CHANGES_REQUESTED** [review]
Overall the patch looks good and things make sense.

I don't understand large parts of it, the ones doing the actual OpenGL rendering, so I basically have to trust you on that. I added a bunch (sorry about that) comments of some other things that should be addressed.

File: .gitmodules (hunk 1 above)
Comment (elima) [reply .gitmodules:6]: Because I had to add [a patch](https://github.com/Igalia/tinygltf/commit/a83211097dacb55742f68b965564170818cf5c6f) for one thing that was missing. Once it is upstream, we can consider switching to the original repo.

File: app/CMakeLists.txt (hunk 2 above)
Comment (elima) [reply app/CMakeLists.txt:231]: Ok

File: app/src/main/cpp/Controller.h (hunk 3 above)
Comment (elima) [reply app/src/main/cpp/Controller.h:49]: Right, I thought about but since I don't haven't use unique_ptr before, I felt conservative :). Will change that.

File: app/src/main/cpp/ControllerDelegate.h (hunk 4 above)
Comment (elima) [reply app/src/main/cpp/ControllerDelegate.h:41]: Yes, this was used for the texture coords attributes, which were not used at the end. Will remove it.

File: app/src/main/cpp/HandModels.cpp (hunk 6 above)
Comment (elima) [reply app/src/main/cpp/HandModels.cpp:157]: Ok

File: app/src/main/cpp/HandModels.cpp (hunk 7 above)
Comment (elima) [reply app/src/main/cpp/HandModels.cpp:167]: Ok

File: app/src/main/cpp/HandModels.cpp (hunk 8 above)
Comment (elima) [reply app/src/main/cpp/HandModels.cpp:177]: I think not showing hands is a better experience than crashing the browser. There are many things that can fail with the GL program, though it is true that not at this point, but in `InitializeGL()`.

File: app/src/main/cpp/HandModels.cpp (hunk 9 above)
Comment (elima) [reply app/src/main/cpp/HandModels.cpp:193]: We are not applying any model-wide transformation because it is not needed. We are animating the mesh via the joints.
This would be useful if we for example wanted to scale the whole hand to make a bit bigger or smaller, or rotate it a bit, etc. But it should not be necessary if the joint locations are accurate.

File: app/src/main/cpp/HandModels.cpp (hunk 11 above)
Comment (elima) [reply app/src/main/cpp/HandModels.cpp:232]: That's a left-over, will remove it.

File: app/src/main/cpp/HandModels.cpp (hunk 12 above)
Comment (elima) [reply app/src/main/cpp/HandModels.cpp:249]: Ok

File: app/src/main/cpp/HandModels.cpp (hunk 13 above)
Comment (elima) [reply app/src/main/cpp/HandModels.cpp:282]: It is not from OpenXR, I declared it at the top of HandModels.cpp. We need to have a maximum allocation for the joints arrays.

File: app/src/main/cpp/BrowserWorld.cpp (hunk 18 above)
Comment (elima) [reply app/src/main/cpp/BrowserWorld.cpp:1155]: For HandModels files, I copied the structure of ExternalBlitter, and it follows the logical life-cycle of a frame (start, draw, end/cancel). I think there is value in leaving this flow explicit, but don't feel strongly about it.

File: app/src/main/cpp/BrowserWorld.cpp (hunk 20 above)
Comment (elima) [reply app/src/main/cpp/BrowserWorld.cpp:1724]: That check is done in `Draw()` method: we bail if `controller.hasHandMesh` is false.

File: app/src/main/cpp/ControllerContainer.cpp (hunk 22 above)
Comment (elima) [reply app/src/main/cpp/ControllerContainer.cpp:240]: We had that before too. It is the way we use to know that we have not initialized the hand joint transforms yet. Only that now it is only done for PicoXR.

File: app/src/main/cpp/ControllerContainer.cpp (hunk 23 above)
Comment (elima) [reply app/src/main/cpp/ControllerContainer.cpp:287]: Right, will fix that.

File: app/src/main/cpp/ControllerContainer.cpp (hunk 24 above)
Comment (elima) [reply app/src/main/cpp/ControllerContainer.cpp:377]: Ok

File: app/src/main/cpp/ControllerContainer.cpp (hunk 25 above)
Comment (elima) [reply app/src/main/cpp/ControllerContainer.cpp:383]: Ok

File: app/src/main/cpp/HandModels.cpp (hunk 17 above)
Comment (elima) [reply app/src/main/cpp/HandModels.cpp:314]: Ok, but again, I'm afraid that in some scenario this could crash the browser, which from a user perspective is much worse than not showing the hands.

File: app/src/main/cpp/ControllerContainer.cpp (hunk 26 above)
Comment (elima) [reply app/src/main/cpp/ControllerContainer.cpp:410]: Agree, but the problem is that there could be more attributes that we shall ignore, so we have to explicitly set the variable to positive. It is a bit more code, but fair enough.

File: app/src/main/cpp/ControllerContainer.cpp (hunk 28 above)
Comment (elima) [reply app/src/main/cpp/ControllerContainer.cpp:254]: He submitted a fix for this in https://github.com/Igalia/wolvic/pull/677 and it is now integrated. The spheres look great now with the lighting.

It is true that this change is not relevant for the hand models, so I can drop it if you want. But IMHO it does improve the look and feel of the joint spheres on Pico.

File: app/src/main/cpp/BrowserWorld.cpp (hunk 18 above)
Comment (elima) [reply app/src/main/cpp/BrowserWorld.cpp:1155]: Do you still prefer to remove them? Either way is fine for me.

File: app/src/main/cpp/ControllerContainer.cpp (hunk 21 above)
Comment (elima) [reply app/src/main/cpp/ControllerContainer.cpp:213]: Unfortunately it is not really a wrong indentation, this block was just taken out of an outer if block that is no longer relevant.

Previously, we were using the lazy construction of the hand joint spheres to also construct the back-nav icon. But now that the spheres are only constructed for Pico, and the back-nav needs to be constructed for all devices except Oculus, we had to rely on other mechanism to perform the lazy-load.

This part of the code is the most confusing one, partly because we had to keep the hand joint spheres code path but only for Pico, and also because we had mingled the back-nav code inside the wrong block :/

File: app/src/main/cpp/BrowserWorld.cpp [hunk 29]
```diff
@@ -531,10 +531,9 @@ BrowserWorld::State::UpdateControllers(bool& aRelayoutWidgets) {
     // incorrect data on some joints; instead, a sphere is drawn for each joint.
 #if !defined(PICOXR)
     // Lazy-load hand models
-    if (controller.mode == ControllerMode::Hand && !controller.hasHandMesh) {
+    if (controller.mode == ControllerMode::Hand && controller.handMesh == nullptr) {
```
Comment (svillar) [inline app/src/main/cpp/BrowserWorld.cpp:534]: better `!controller.handMesh`

File: app/src/main/cpp/BrowserWorld.cpp [hunk 30]
```diff
@@ -1719,7 +1718,7 @@ BrowserWorld::DrawWorld(device::Eye aEye) {
 
   // Draw hand models
   for (Controller& controller: m.controllers->GetControllers()) {
-    if (controller.enabled && controller.mode == ControllerMode::Hand)
+    if (controller.enabled && controller.mode == ControllerMode::Hand && controller.handMesh != nullptr)
```
Comment (svillar) [inline app/src/main/cpp/BrowserWorld.cpp:1721]: Same, enough with `controller.handMesh`

File: app/src/main/cpp/Controller.cpp [hunk 31]
```diff
@@ -77,8 +77,8 @@ Controller::operator=(const Controller& aController) {
   batteryLevel = aController.batteryLevel;
   handMeshToggle = aController.handMeshToggle;
   handJointTransforms = aController.handJointTransforms;
-  memcpy (&handMesh, &aController.handMesh, sizeof(handMesh));
-  hasHandMesh = aController.hasHandMesh;
+  if (aController.handMesh != nullptr)
```
Comment (svillar) [inline app/src/main/cpp/Controller.cpp:80]: Ditto

File: app/src/main/cpp/HandModels.h (hunk 5 above)
Comment (svillar) [reply app/src/main/cpp/HandModels.h:7]: Yes we actually use it quite a lot

File: app/src/main/cpp/ControllerContainer.cpp (hunk 28 above)
Comment (svillar) [reply app/src/main/cpp/ControllerContainer.cpp:254]: OK, let's keep it then

**Review by svillar — CHANGES_REQUESTED** [review]
Just a couple of nits

File: app/src/main/cpp/HandModels.h (hunk 5 above)
Comment (elima) [reply app/src/main/cpp/HandModels.h:7]: Ok, will fix that.

File: app/src/main/cpp/BrowserWorld.cpp (hunk 29 above)
Comment (elima) [reply app/src/main/cpp/BrowserWorld.cpp:534]: Ok

**Review by svillar — APPROVED** [review]
Thanks, we're ready to merge it I believe.

Mind merging the last two commits addressing the review comments in the other ones?

**Comment by elima** [comment]
> Mind merging the last two commits addressing the review comments in the other ones?

Will do, thanks!

File: app/src/main/cpp/BrowserWorld.cpp (hunk 18 above)
Comment (svillar) [reply app/src/main/cpp/BrowserWorld.cpp:1165]: Yeah I think we can remove them because we don't have a clear use case for them. We can always add them later
